### PR TITLE
fix(storage): make Query show Prefix consistently

### LINKED ISSUE #1763: storage: query.SetAttrSelection hide prefixes if a Delimiter is specified
**Client**

storage

**Environment**

Linux

**Code**

```go
package main

import (
	"context"
	"fmt"

	"cloud.google.com/go/storage"
	"google.golang.org/api/iterator"
)

const (
	bucketName = "set your bucket name here"
)

var (
	gcsFieldsSelection = []string{"Bucket", "Name", "ContentType", "ContentLanguage", "CacheControl", "EventBasedHold",
		"TemporaryHold", "RetentionExpirationTime", "ACL", "Owner", "ContentEncoding", "ContentDisposition", "Size", "MD5",
		"CRC32C", "MediaLink", "Metadata", "Generation", "Metageneration", "StorageClass", "CustomerKeySHA256", "KMSKeyName",
		"Created", "Deleted", "Updated", "Etag"}
)

func main() {
	ctx := context.Background()
	svc, err := storage.NewClient(ctx)
	if err != nil {
		fmt.Printf("error creating storage client: %v\n", err)
	}
	query := &storage.Query{Prefix: "", Delimiter: "/"}
	bkt := svc.Bucket(bucketName)
	it := bkt.Objects(ctx, query)
	for {
		attrs, err := it.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			fmt.Printf("error listing bucket contents with no attribute selection: %v\n", err)
			break
		}
		if len(attrs.Prefix) > 0 {
			fmt.Printf("attr prefix no attr selection: %v\n", attrs.Prefix)
		} else {
			fmt.Printf("attr name no attr selection: %v\n", attrs.Name)
		}
	}
	fmt.Println()
	fmt.Printf("now the same query with attr selection, the same results are expected\n")
	fmt.Println()
	query = &storage.Query{Prefix: "", Delimiter: "/"}
	err = query.SetAttrSelection(gcsFieldsSelection)
	if err != nil {
		fmt.Printf("error setting attribute selection: %v\n", err)
		return
	}
	it = bkt.Objects(ctx, query)
	for {
		attrs, err := it.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			fmt.Printf("error listing bucket contents with attribute selection: %v\n", err)
			break
		}
		if len(attrs.Prefix) > 0 {
			fmt.Printf("attrs prefix with attr selection: %v\n", attrs.Prefix)
		} else {
			fmt.Printf("attrs name with attr selection: %v\n", attrs.Name)
		}
	}
}
```

**Expected behavior**

`bkt.Objects` should return the same number or results with and without `SetAttrSelection`

**Actual behavior**

when setting `SetAttrSelection` the "folders" are not listed, here is the output of the provided test code for my bucket

```
attr name no attr selection: ros.txt
attr name no attr selection: sftpgo.log
attr prefix no attr selection: testdir/
attr prefix no attr selection: vdir/

now the same query with attr selection, the same results are expected

attrs name with attr selection: ros.txt
attrs name with attr selection: sftpgo.log
```

so the folders `testdir/` and `vdir/` are returned only without `SetAttrSelection`

### PR DESCRIPTION
**Pull request by tritone**
A Query with SetAttrSelection currently hides Prefix results,
which are returned otherwise. This is a fix to make sure
Prefixes are not inadvertently filtered.

Fixes #1763.

Credit to @diggerk for the fix (originally in https://code-review.googlesource.com/c/gocloud/+/56712/1/storage/storage.go)

### CODE REVIEW

File: storage/storage.go [hunk 1]
```diff
@@ -1362,7 +1362,7 @@ func (q *Query) SetAttrSelection(attrs []string) error {
 
 	if len(fieldSet) > 0 {
 		var b bytes.Buffer
-		b.WriteString("items(")
+		b.WriteString("prefixes,items(")
```
Comment (frankyn) [inline storage/storage.go:1365]: Could you add a verification in the system tests for SetAttrSelection?

**Comment by codyoss** [comment]
Might be worth auditing our other languages at some point to see if they have this issue as well. LGTM.

**Comment by mstfymrtc** [comment]
Is there any progress on this?

File: storage/storage.go (hunk 1 above)
Comment (tritone) [reply storage/storage.go:1389]: Done. Really should convert the whole of TestIntegration_Objects to use table-driven style and subtests, but another time :)

File: storage/integration_test.go [hunk 2]
```diff
@@ -686,6 +686,43 @@ func TestIntegration_Objects(t *testing.T) {
...
+		var gotNames []string
+		var gotPrefixes []string
+		it := bkt.Objects(context.Background(), query)
+		for {
+			attrs, err := it.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				log.Fatal(err)
```
Comment (codyoss) [inline storage/integration_test.go:704]: Should this be a t.Fatal?

Comment (tritone) [reply storage/integration_test.go:704]: Good catch, yes.

Comment (tritone) [reply storage/integration_test.go]: Fixed in the other places which use this loop as well.

**Review by codyoss — APPROVED** [review]
LGTM

File: storage/integration_test.go [hunk 3]
```diff
@@ -686,6 +686,43 @@ func TestIntegration_Objects(t *testing.T) {
...
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				t.Fatalf("iterator.Next: %v", err)
+			}
+			if attrs.Name != "" {
+				gotNames = append(gotNames, attrs.Name)
+			} else if attrs.Prefix != "" {
+				gotPrefixes = append(gotPrefixes, attrs.Prefix)
```
Comment (frankyn) [inline storage/integration_test.go:709]: Only a comment: a lot cleaner

File: storage/integration_test.go [hunk 4]
```diff
@@ -686,6 +686,43 @@ func TestIntegration_Objects(t *testing.T) {
...
+			}
+
+			if len(attrs.Bucket) > 0 {
```
Comment (frankyn) [inline storage/integration_test.go:712]: Suggestion, but otherwise LGTM:

Is there a preference over this vs. what's used above `if attrs.Bucket != ""`?

**Review by frankyn — APPROVED** [review]

File: storage/integration_test.go (hunk 4 above)
Comment (tritone) [reply storage/integration_test.go:712]: Ah, I think != "" seems more typical but I had just copied this over from the existing test for SetAttrSelection.

Comment (frankyn) [reply storage/integration_test.go:712]: Not sure how len() is handled but might be a comparison between O(n) and O(1).

**Pull request merged**